### PR TITLE
Issue #1669: Adding extension methods ConfigureMvcOptions and ConfigureRouteOptions.

### PR DESCRIPTION
**Pull request by sornaks**
Linking https://github.com/aspnet/Mvc/issues/1669

### CODE REVIEW

File: src/Microsoft.AspNet.Mvc/MvcServiceCollectionExtensions.cs [hunk 1]
```diff
@@ -20,6 +17,28 @@ public static IServiceCollection AddMvc(this IServiceCollection services, IConfi
             return services;
         }
 
+        /// <summary>
+        /// Configures a set of MvcOptions for the application.
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc/MvcServiceCollectionExtensions.cs]: use `<see cref="MvcOptions" />` where you're using a class name

File: src/Microsoft.AspNet.Mvc/MvcServiceCollectionExtensions.cs [hunk 2]
```diff
@@ -20,6 +17,28 @@ public static IServiceCollection AddMvc(this IServiceCollection services, IConfi
...
+            services.Configure(setupAction);
+        }
+
+        /// <summary>
+        /// Configures a set of RouteOptions for the application.
+        /// </summary>
+        /// <param name="services">The services available in the application.</param>
+        /// <param name="setupAction">The RouteOptions which need to be configured.</param>
+        public static void ConfigureRouteOptions(
+            this IServiceCollection services, [NotNull]Action<RouteOptions> setupAction)
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc/MvcServiceCollectionExtensions.cs]: when you break up parameters, put them each on their own line.

space between `[NotNull]` and the rest

File: src/Microsoft.AspNet.Mvc/MvcServiceCollectionExtensions.cs [hunk 3]
```diff
@@ -20,6 +17,28 @@ public static IServiceCollection AddMvc(this IServiceCollection services, IConfi
...
+        {
+            services.Configure(setupAction);
+        }
+
+        /// <summary>
+        /// Configures a set of RouteOptions for the application.
+        /// </summary>
+        /// <param name="services">The services available in the application.</param>
+        /// <param name="setupAction">The RouteOptions which need to be configured.</param>
+        public static void ConfigureRouteOptions(
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc/MvcServiceCollectionExtensions.cs]: Why is this defined in MVC? This really seems like it should be defined in routing.

File: test/WebSites/BasicWebSite/Filters/GlobalActionFilter.cs [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+                    Content = "This was added by filter.",
+                    ContentType = "text/plain"
+                };
+            }
+        }
+
+        public void OnActionExecuting(ActionExecutingContext context)
+        {
+        }
+    }
```
Comment (rynowak) [inline test/WebSites/BasicWebSite/Filters/GlobalActionFilter.cs]: Why not use ApplicationModel to add this to the action that you're interested in. These 'global' filters that only act on one action are a bad example for users and are hard to debug.

**Comment by sornaks** [comment]
Updated. Moved ConfigureRouteOptions to aspnet/Routing repo. PR - https://github.com/aspnet/Routing/pull/146

File: test/Microsoft.AspNet.Mvc.FunctionalTests/BasicTests.cs [hunk 5]
```diff
@@ -259,5 +259,21 @@ public static IEnumerable<object[]> HtmlHelperLinkGenerationData
...
+
+        [Fact]
+        public async Task ConfigureMvcOptionsAddsOptionsProperly()
+        {
+            // Arrange
+            var server = TestServer.Create(_provider, _app);
+            var client = new HttpClient(server.CreateHandler(), false);
+
+            // Act
+            var response = await client.GetAsync("http://localhost/Home/GetApplicaitonDescription");
```
Comment (yishaigalatzer) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/BasicTests.cs]: Typo : GetApplicationDescription

File: test/WebSites/BasicWebSite/Controllers/HomeController.cs [hunk 6]
```diff
@@ -57,5 +57,11 @@ public IActionResult JsonTextInView()
         {
             return View();
         }
+        
+        public string GetApplicaitonDescription()
```
Comment (yishaigalatzer) [inline test/WebSites/BasicWebSite/Controllers/HomeController.cs]: `GetApplicationDescription`

**Comment by yishaigalatzer** [comment]
Also change the samples:
Here: https://github.com/aspnet/Mvc/blob/dev/samples/MvcSample.Web/Startup.cs#L63

Add an extension method for: `RazorViewEngineOptions`

And look for other Option classes we have, and lets discuss if we want extension methods for them as well.

**Comment by sornaks** [comment]
We have Option classes for Authorization for which we already have an extension method.

Updated the PR with `ConfigureRazorViewEngineOptions`. Please have a look.

File: src/Microsoft.AspNet.Mvc.Razor/RazorServiceCollectionExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,26 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.Framework.DependencyInjection;
+
+namespace Microsoft.AspNet.Mvc.Razor
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Razor/RazorServiceCollectionExtensions.cs]: This doesn't seem like the namespace we want

File: src/Microsoft.AspNet.Mvc/MvcServiceCollectionExtensions.cs [hunk 8]
```diff
@@ -20,6 +17,18 @@ public static IServiceCollection AddMvc(this IServiceCollection services, IConfi
...
         }
 
+        /// <summary>
+        /// Configures a set of <see cref="MvcOptions"/> for the application.
+        /// </summary>
+        /// <param name="services">The services available in the application.</param>
+        /// <param name="setupAction">The <see cref="MvcOptions"/> which need to be configured.</param>
+        public static void ConfigureMvcOptions(
+            this IServiceCollection services,
+            [NotNull] Action<MvcOptions> setupAction)
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc/MvcServiceCollectionExtensions.cs]: should be notnull on both

**Comment by rynowak** [comment]
:shipit: from me after fixing namespace

**Comment by sornaks** [comment]
Thanks! Checked in - https://github.com/aspnet/Mvc/commit/6e8cc6ba74089f5d821c2ae8342d7e8ec50448fc

**Pull request closed without merging**